pkg/sharding/handler.go

```diff
@@ -0,0 +1,202 @@
+		if limit := req.URL.Query().Get("limit"); limit == "" {
+			http.Error(w, "sharded LIST calls require paging", http.StatusInternalServerError)
+			return
+		}
+
+		state, err := NewState(req.URL.Query(), shards, shardResourceVersion)
+		if err != nil {
+			klog.Warningf("error parsing sharded chunked state: %v", err)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		shard, continueToken, err := state.NextQuery()
+		if err != nil {
+			klog.Warningf("error determining next sharded chunked query: %v", err)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		loader.RLock()
+		client := loader.Clients[shard]
+		loader.RUnlock()
+		resp, err := delegateTo(client, continueToken, req)
+		if err != nil {
+			klog.Warningf("error sending sharded request: %v", err)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		if resp.StatusCode != http.StatusOK {
+			copyInto(w, resp.Header, resp.StatusCode, resp.Body)
+			return
+		}
+		// update vector clock, resource version in data, send
+		raw, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			klog.Warningf("error reading sharded request body: %v", err)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		if err := resp.Body.Close(); err != nil {
+			klog.Warningf("error closing sharded request body: %v", err)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		obj, _, err := unstructured.UnstructuredJSONScheme.Decode(raw, nil, nil)
+		if err != nil {
+			klog.Warningf("could not parse sharded request body as unstructured: %v", err)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		switch t := obj.(type) {
+		case *unstructured.UnstructuredList:
+			if err := state.UpdateWith(shard, t); err != nil {
+				klog.Warningf("could not update sharded state: %v", err)
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+			updatedResourceVersion, err := state.ToResourceVersions().Encode()
+			if err != nil {
+				klog.Warningf("could not format new resource version: %v", err)
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+			t.SetResourceVersion(updatedResourceVersion)
+			updatedContinueToken, err := state.Encode()
+			if err != nil {
+				klog.Warningf("could not format new continue token: %v", err)
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+			t.SetContinue(updatedContinueToken)
+			for _, item := range t.Items {
+				item.SetClusterName(fmt.Sprintf("%s/clusters/%s", shard, item.GetClusterName()))
+			}
+			body := bytes.Buffer{}
+			if err := unstructured.UnstructuredJSONScheme.Encode(t, &body); err != nil {
+				klog.Warningf("could not encode sharded response body: %v", err)
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+			resp.Header.Set("Content-Length", strconv.Itoa(body.Len()))
```

Comment:
This shouldn't be necessary, it should already be set to the length of the contents.